callable/src/main/java/io/grpc/callable/Callable.java

```diff
@@ -0,0 +1,423 @@
+   * operations applied to it may deadlock because of this. However, it is safe to use this
+   * callable in the context of a {@link #followedBy(Callable)} operation, which is the major
+   * use cases for transformers. But if you use a transformer to simulate a real rpc
+   * you should use {@link #create(Transformer, Executor)} instead.
+   */
+  public static <RequestT, ResponseT> Callable<RequestT, ResponseT>
+      create(Transformer<? super RequestT, ? extends ResponseT> transformer) {
+    return new TransformingCallable<RequestT, ResponseT>(transformer, null);
+  }
+
+  /**
+   * Interface for a transformer. It can throw a {@link StatusException} to indicate an error.
+   */
+  public interface Transformer<RequestT, ResponseT> {
+    ResponseT apply(RequestT request) throws StatusException;
+  }
+
+  /**
+   * Returns a callable which echos its input.
+   */
+  public static <RequestT> Callable<RequestT, RequestT>
+      identity() {
+    return new TransformingCallable<RequestT, RequestT>(new Transformer<RequestT, RequestT>() {
+      @Override public RequestT apply(RequestT request) throws StatusException {
+        return request;
+      }
+    }, null);
+  }
+
+  /**
+   * Returns a callable which always returns the given constant.
+   */
+  public static <RequestT, ResponseT> Callable<RequestT, ResponseT>
+      constant(final ResponseT result) {
+    return new TransformingCallable<RequestT, ResponseT>(new Transformer<RequestT, ResponseT>() {
+      @Override public ResponseT apply(RequestT request) throws StatusException {
+        return result;
+      }
+    }, null);
+  }
+
+  // Followed-By
+  // ===========
+
+  /**
+   * Returns a callable which forwards the responses from this callable as requests into the other
+   * callable. Works both for unary and streaming operands. Example:
+   *
+   * <pre>
+   * String bookName = ...;
+   * Callable.Transformer&lt;Book, GetAuthorRequest> bookToGetAuthorRequest = ...;
+   * Author response =
+   *     Callable.create(LibraryGrpc.CONFIG.getBook)
+   *             .followedBy(Callable.create(bookToGetAuthorRequest))
+   *             .followedBy(Callable.create(LibraryGrpc.CONFIG.getAuthor))
+   *             .call(channel, new GetBookRequest().setName(bookName).build());
+   * </pre>
+   *
+   * <p>For streaming calls, each output of the first callable will be forwarded to the second
+   * one as it arrives, allowing for streaming pipelines.
+   */
+  public <FinalResT> Callable<RequestT, FinalResT>
+      followedBy(Callable<ResponseT, FinalResT> callable) {
+    return new FollowedByCallable<RequestT, ResponseT, FinalResT>(this, callable);
+  }
+
+  // Channel Redirection
+  // ===================
+
+  /**
+   * Returns a callable which redirects the execution of this callable to a channel determined at
+   * call execution time, via a supplier. This can be used for creating batches which target
+   * different services. Example:
+   *
+   * <pre>
+   * Author response =
```

Comment:
What is this used for? Load balancing or is it an affordance for DI so that everything else can be wired up in advance except the Channel.